Chat message: Quote code

estib-vega opened the pull request:
This introduces the ability to select some lines from the patch diff hunks, and quote it in a message.

Introduces:
- HunkDiff selection handling (click toggles and drag selection)
- Chati input quote preview (display what part of the diff is being quoted)
- Propagate the diff selection to the parametes for sending a message
- Parse out the diff quote attributes from the message and display them

viz:

https://github.com/user-attachments/assets/68e3e5a6-cc9f-4d6c-ac22-5a69dc5b20e6


<!-- GitButler Footer Boundary Top -->
---
This is **part 2 of 2 in a stack** made with GitButler:
- <kbd>&nbsp;2&nbsp;</kbd> #7279 👈 
- <kbd>&nbsp;1&nbsp;</kbd> #7274 
<!-- GitButler Footer Boundary Bottom -->

estib-vega committed:
Hunk diff: Line selection

Add some properties for externally handling the line selection of the diff whenever the user clicks on a line.

Current selection stye for that is not vetted

estib-vega committed:
Review: Diff line selection

Handle the selection of diff lines for chat/comment purposes

estib-vega committed:
Display the code selection in the chat editor

Display the current code selected in the chat input field

estib-vega committed:
Chat: Propagate the diff selection

Send the diff selection parameters if any

estib-vega committed:
Factor out hunk diff body 

Create a stand-alone component called `HunkDiffBody` in order to be able to reuse the syntax highlighting of the diff patches quoted on messages

estib-vega committed:
Chat Message: Display quoted text

Display the quoted diff patch content in the chat message, if any

estib-vega commented:
@PavelLaptev Sorry for the late tagging

The code changes are done from my side, there are only two main things that need to be style-vetted:

### 1: The selection state in the hunk diff

<img width="639" alt="Screenshot 2025-02-10 at 20 07 00" src="https://github.com/user-attachments/assets/cbcb56ec-c006-49e5-8484-cd34347e3947" />

### 2: The 'preview' of the quote.
This is shown in the input field to signal that there is code that will be quoted in the message that's about to be sent

<img width="601" alt="Screenshot 2025-02-10 at 20 07 33" src="https://github.com/user-attachments/assets/d15b1007-68e4-4a44-822e-1b0753aa405a" />

PavelLaptev commented:
Thank you for the PR, @estib-vega!  

I've added the code quotation design [here in Figma](https://www.figma.com/design/hrEqnFaxS6mDPBqKfIOOZg/Web-Client?node-id=1418-36223&t=6dWB839UXlFoTxdZ-1).  

#### Key edits:  

1. Rows are selected only when clicking on the count columns, not the entire row.  
2. The code quote box appears in the dialog only when clicking the **"quote"** button. The second button is just for **copying** the code.  

   ![image](https://github.com/user-attachments/assets/15398c53-bc66-4064-9507-ec42bf95475e)  

3. Selection should be indicated with an overlay box.  
4. Selection should support using the **Shift** key.  
5. It would be nice to support selection using **keyboard arrows**.  
6. Code quotes should be removable from the dialog box using a **"cross"** button.  

   ![image](https://github.com/user-attachments/assets/de24e425-d531-4504-885a-07b8f276c82a)  

---

Let me know if you have any further questions! 🚀

estib-vega commented:
@PavelLaptev 
Thanks for the review! I'll do the changes 👍

estib-vega committed:
Update the line selection style

estib-vega committed:
Line selection only triggered from number column

estib-vega committed:
Line selection preview has a close button

Add a 'cross' button to clear the selection

estib-vega committed:
Line selection: Keep track of group hierarchy 

Keep track of whether a line is the first or last (or both) in a selection group

estib-vega committed:
Overflow menu for quote and copy of code

Display an overflow menu over the code selection when hovering over the table body.

This displays the buttons for quoting the code on the chat message and to copy the lines
